[로또] 코맥 미션 제출합니다.

parker1609 opened the pull request:
감사합니다.

parker1609 committed:
로또 1단계 요구사항 분석

parker1609 committed:
feat: 로또 번호 생성

로또 번호 객체는 숫자 범위 1 ~ 45를 가진다.

parker1609 committed:
refactor: 로또 번호 범위 상수화

* 로또 번호 범위를 나타내는 매직넘버를 static final로 상수화한다.

parker1609 committed:
feat: 로또 생성

* 정상적인 로또 번호 객체 6개를 저장하는 로또 객체

* 생성자 인자에서 중복된 번호가 있는 경우 예외 처리한다.

parker1609 committed:
refactor: 로또 번호 개수 상수화

parker1609 committed:
chore: 추 후에 적용할 Todo 항목 추가

parker1609 committed:
feat: 로또 예외 테스트 코드 추가

parker1609 committed:
feat: 랜덤 번호 생성

* RandomNumberGenerator.java: 1 ~ 45 숫자를 저장하고 있는 리스트에서
shuffle()을 통해 6개의 무작위 숫자를 뽑는다.

parker1609 committed:
refactor: 상수 enum으로 추출

* 로또를 만드는 규칙을 따로 LottoRule enum으로 추출하였다.

parker1609 committed:
refactor: 로또 금액 객체 생성

* 로또 금액을 담당하는 Money 객체 생성

* 최소 금액보다 적은 금액으로 Money 객체를 만드는 예외 처리

* 최소 금액 테스트 코드 구현

parker1609 committed:
feat: 구매할 수 있는 로또 개수 계산

* Money.java: getNumberOfLotto() 메서드는 구매할 수 있는 로또 개수를
계산한다.

* 계산 결과가 정확한지 테스트 코드를 구현하였다.

parker1609 committed:
refactor: 매직넘버 상수로 변경

parker1609 committed:
feat: 로또 번호 일치 개수 계산 구현

* WinningLotto.java: 당첨 번호를 관리하는 객체

* Lotto.java: 2개의 로또 중에서 번호가 일치하는 계수를 계산하는
calculateCountOfMatch() 메서드 구현

* 로또 번호 일치 개수 계산 테스트 코드 구현

parker1609 committed:
refactor: 로또 등수 Rank enum 객체로 추출

* WinningLotto.java: getWinning() 메서드가 로또 등수를 나타내는 Rank
객체를 반환하도록 변경하였다.

parker1609 committed:
README: 당첨 결과에 대한 요구사항 수정

parker1609 committed:
feat: 로또 금액 입력 후 생성된 로또 출력 구현

* Main.java: 로또 금액을 입력받은 후 그에 따라 로또 개수만큼 로또를 생성한다.

* InputView.java: 로또 구입 금액을 입력받는다.

* OutputView.java: 생성한 로또를 출력한다.

* Lotto.java, LottoNumber.java: 로또 출력을 위한 메서드를 추가하였다.

parker1609 committed:
refactor: List<Lotto>를 Lottos 객체로 추출

parker1609 committed:
refactor: 금액 입력 오류시 재입력 받기

parker1609 committed:
feat: 지난 주 당첨 번호 입력 받기

parker1609 committed:
refactor: 지난 주 당첨 번호 입력 오류시 재입력 받기

parker1609 committed:
refactor: 불필요한 변수 제거 및 view 로직 옮기기

* Main.java: 금액과 지난 로또 당첨 번호를 만드는 메서드에서 불필요한
메서드를 제거하였다.

* Money.java: Money 객체에서 view 로직을 옮기기 위해 인자를 String에서
int로 변경하였다.

* InputView.java: String을 입력받아 integer로 변경하는 기능을 추가하였다.

parker1609 committed:
refactor: 클래스 이름 변경

parker1609 committed:
refactor: 로또 번호 일치 결과 구현

* LottoResult.java: 로또 결과를 책임지는 객체이고, getRank()메서드는
전체 로또 번호 일치 결과를 반환한다.

* LottoResultTest.java: 로또 번호 일치가 정상적으로 동작하는지 테스트 코드를 구현하였다.

parker1609 committed:
refactor: 현재 불필요한 Rank 삭제

parker1609 committed:
feat: 총 수익률 계산

* LottoResult.java: getEarningsRate() 메서드는 총 수익률을 계산한다.
getEarning() 메서드는 총 수익을 계산한다.

* LottoResultTest.java: 총 수익률이 정상적으로 계산되는지 테스트 코드를
구현하였다.

parker1609 committed:
refactor: 상수화 및 LottoRule로 추출

parker1609 committed:
refactor: 로또 등수 결과 로직 변경

* LottoResult.java: 로또 등수를 세기 위해 각 등수와 해당 등수에 당첨된
횟수를 저장하는 Map 자료구조를 사용하였다.

parker1609 committed:
feat: 당첨 통계 출력

parker1609 committed:
refactor: 자료구조 변경

* Lotto.java: 로또 번호를 저장하는 lottoNumbers 자료구조를 Set에서
List로 변경하였다. 이 이유는 sort를 간편하게 하기 위함이다. 중복을
검사하는 로직은 따로 checkDuplication() 메서드로 분리하였다.

* LottoNumber.java: 로또 번호를 오름차순으로 하기 위해 compareTo()
메서드를 추가하였다.

parker1609 committed:
refactor: 공백 조정

parker1609 committed:
refactor: 인스턴스 미리 생성

* LottoNumber.java: 로또 번호 1 ~ 45를 미리 생성하여 Map에 저장해두고,
메서드를 통해 꺼내 사용하도록 구현하였다.

parker1609 committed:
feat: 보너스 번호 추가

* WinningLotto.java: 로또 결과에 보너스 번호가 일치하는지 판단하는
로직을 추가한다.

* Rank.java: 보너스 볼이 일치하면 2등으로 판단한다.

* Lotto.java: LottoNumber 리스트 안에 특정 LottoNumber가 있는지 판단하는
isContain() 메서드를 추가하였다.

parker1609 committed:
refactor: 테스트 코드 추가

parker1609 committed:
refactor: 보너스 번호 입력 추가

* WinningLotto.java: 보너스 번호가 지난 주 당첨 로또 번호과 중복되는
것이 있는지 검사하는 기능을 구현하였다.

* Main.java: WinningLotto 객체를 정상적으로 생성하는
createWinningLotto() 메서드를 구현하였다.

* InputConsole.java: 보너스 번호 입력하는 inputBonusNumber() 메서드를
구현하였다.

parker1609 committed:
refactor: 사용하지 않는 import문 제거

parker1609 committed:
feat: 2등 엔트리 및 출력 추가

* LottoResult.java: Map에 2등 엔트리를 추가하였다.

* OutputConsole.java: 2등 출력을 추가하였다.

parker1609 committed:
README: 3단계 요구 사항 분석

parker1609 committed:
feat: 수동 로또 구매 기능 추가

* Main.java: 수동 로또 구매 개수 및 로또 번호를 생성한다.

* InputConsole.java: 수동 로또와 지난 당첨 로또는 동일하므로 이를
inputLotto() 메서드로 통일한다. 수동 로또 구매 개수를 입력하는
inputNumberOfManualLotto() 메서드를 추가하였다.

* OutputConsole.java: 수동 로또 출력을 추가하였다.

jihan805 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,50 @@
+    public int calculateCountOfMatch(Lotto anotherLotto) {
+        int countOfMatch = 0;
+        for (LottoNumber lottoNumber : this.lottoNumbers) {
+            countOfMatch += anotherLotto.lottoNumbers.contains(lottoNumber) ? 1 : 0;
```
아래 구현한 isContain 를 사용하면 어떨까요?

parker1609 replied:
피드백 감사합니다! 수정하였습니다.

jihan805 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,56 @@
+        this.number = number;
+    }
+
+    public static LottoNumber get(int number) {
```
재사용 가능한 구조 적용 💯 
LottoNumber 의 인스턴스를 get한다는 의미로 get 보다는 getInstance 와 같이 조금 더 명확한 의미로 사용하면 좋을 것 같아요!

parker1609 replied:
피드백 감사합니다! getInstance()로 변경하였습니다.

jihan805 commented on src/main/java/lotto/domain/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,22 @@
+public class RandomNumberGenerator {
+    private static final int START = 0;
+
+    private static List<Integer> allLottoNumbers = new ArrayList<>();
```
Integer대신 LottoNumber를 사용하면 어떨까요?

parker1609 replied:
피드백 감사합니다.

LottoNumber 객체를 만든 이유는 Integer값을 포장하여 정상적인 객체임을 보장한 후, 이를 객체 사이에서 사용하기 위함이라고 이해하였습니다. RandomNumberGenerator.java에서 Integer를 LottoNumber 객체로 변경하면 Lotto 객체의 생성자 역시 ```List<LottoNumber>``` 로 매개변수를 받도록 변경하여야 했습니다. 그 결과 원시값을 포장한 LottoNumber 객체를 더욱 잘 활용할 수 있었다고 생각하였습니다.

jihan805 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,51 @@
+    FOURTH(4, 50_000),
+    FIFTH(3, 5_000),
+    MISS(0, 0);
+
```
Rank를 enum 이용해 구현 👍 
보너스볼 여부를 판단하기 위해 bonus인지 아닌지를 판단하는 값을 추가해보면 어떨까요?

parker1609 replied:
피드백 감사합니다.

Rank enum에 보너스볼이 일치하는 지 여부를 추가하였습니다. 그 결과, 매개변수로 전달된 데이터를 통해 어떤 rank인지 판단하는 valueOf() 메서드가 더 이해하기 좋은 코드가 되었다고 생각합니다. 보너스볼 여부가 없을 때는 SECOND를 따로 생각해주어야 했지만, 이를 추가한 결과 모든 rank에 공통적인 로직으로 판단할 수 있었습니다.

parker1609 replied:
리뷰어님 질문있습니다.
저는 보너스볼 여부를 판단하기 위해 FIRST, SECOND 등의 3번째 멤버로 보너스볼인지 아닌지를 판단할 boolean 변수를 추가하였습니다. 그리고 SECOND에는 true값, 그 외의 랭크에는 false값으로 설정하였습니다. 하지만 프로그램을 수행해보면서 예외를 발견하게 되었습니다. 예를 들어 5등인 경우 3개의 로또 번호가 일치하면 보너스볼 여부와 상관없이 5등으로 판단해야 하는데, 이와 같이 boolean 값을 추가한 경우 반드시 보너스볼 번호와 일치하는 번호가 없는 경우에만 5등으로 판단하였습니다. 이를 해결하기 사전에 2등을 처리하고 그 외의 Rank에는 보너스볼이 일치하는지 여부를 나타내는 변수를 전혀 사용하지 않았습니다. 하지만 이는 위에서 boolean변수를 사용하지 않았을 때와 같은 코드가 되었습니다. 이와 같이 변경하는 것이 맞는지 궁금합니다!

jihan805 replied:
2등을 처리하는 대신, matchedBonus가 true 인 경우에만 boolean으로 넘어온 변수에 대해 확인하도록 해도 좋을 것 같아요!

jihan805 commented on src/test/java/lotto/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,15 @@
+        assertThrows(IllegalArgumentException.class, () -> {
+            LottoNumber.get(0);
+        });
+    }
```
경계값을 이용해 테스트 👍 
45 기준의 테스트도 추가해보면 어떨까요?

parker1609 replied:
피드백 감사합니다! 45를 넘어선 번호로 로또를 만들 때 오류가 정상적으로 발생하는지 테스트하는 코드를 추가하였고, 그에 맞는 테스트 메서드 이름으로 변경하였습니다.

jihan805 commented on src/test/java/lotto/domain/MoneyTest.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @Test
+    void 구매할_수_있는_로또_개수_계산2() {
```
다양한 테스트 케이스 구현 👍 
이름 뒤에 2 보다는 구매할 수 있는 로또의 개수인데 어떤 특정한 경우인지를 나타내도록 테스트 메서드의 이름을 작성하면 좋을 것 같아요!

parker1609 replied:
피드백 감사합니다! 테스트 메서드 이름으로 무엇을 테스트하였는지 나타나도록 변경하였습니다.

jihan805 commented on src/main/java/lotto/view/OutputConsole.java:
```diff
@@ -0,0 +1,34 @@
+    public static void outputResult(WinningLotto winningLotto, Lottos lottos) {
+        System.out.println("\n당첨 통계\n----------");
+        LottoResult lottoResult = new LottoResult(winningLotto, lottos);
+        System.out.println("3개 일치 (5,000원) - " + lottoResult.getCountOfRank(FIFTH) + "개");
```
OutputConsole에서 LottoResult를 만들고 있어요!
outputResult(LottoResult lottoResult)와 같이 분리하여 구현해보면 어떨까요?

parker1609 replied:
피드백 감사합니다!

View에서 LottoResult를 만드는 것은 맞지 않다고 이해하여 controller를 담당하는 Main에서 LottoResult 객체를 생성하여 매개변수로 넘겨주도록 변경하였습니다.

jihan805 commented on src/main/java/lotto/Main.java:
```diff
@@ -0,0 +1,67 @@
+        for (int i = 0; i < numberOfManualLotto; i++) {
+            manualLottos.add(createLotto());
+        }
+        return manualLottos;
```
현재 Main에서 로또 생성 로직 일부를 담당하고 있어요!
로또 생성은 자동/수동 방식이 있어요.
둘다 로또를 생성한다는 공통점이 있는데 인터페이스로 충분히 추출할 수 있지 않을까요?

parker1609 replied:
피드백 감사합니다!

수동 로또 생성은 ManualLottoMachine() 메서드, 자동 로또 생성은 AutoLottoMachine() 메서드가 당담하도록 하였습니다. 두 메서드 모두 로또 생성이라는 공통점이 있기 때문에 generateLotto()메서드를 가지고 있습니다. 이를 LottoMachine 인터페이스로 추출하였습니다.

jihan805 reviewed (changes requested):
안녕하세요 코맥님, 
전체적으로 정말 깔끔하게 잘 구현하셨어요 💯 
몇가지 피드백 남겼으니 확인 부탁드려요 :)

parker1609 committed:
refactor: 구현된 함수 사용

parker1609 committed:
refacotr: 명확한 의미를 나타내는 이름으로 변경

* LottoNumber.java: 인스턴스를 생성하는 get() 메서드의 이름을 더 명확한
이름인 getInstance() 메서드로 변경하였다.

parker1609 committed:
refacotr: 테스트 코드 추가

* LottoNumberTest.java: 최대 범위보다 큰 로또 번호를 만들 때 예외가
발생하는지 테스트하는 코드를 추가하였다. 테스트 메서드 이름을 좀 더
명확하게 변경하였다.

parker1609 committed:
refactor: 좀 더 명확한 메서드 이름으로 변경

parker1609 committed:
refactor: 생성자 분리

* Main.java, OutputConsole.java: 콘솔 출력을 담당하는
OutputConsole.java에서 LottoResult를 생성하는 것은 적절하지 않다고
판단하여 Main.java에서 생성하여 이를 매개변수로 전달하였다.

parker1609 committed:
refactor: 로또 생성 객체 분리

* LottoMachine.java: 로또 생성 인터페이스

* AutoLottoMachine.java: 자동으로 로또를 생성하는 객체

* ManualLottoMachine.java: 수동으로 로또를 생성하는 객체

* Main.java: 로또 생성을 수행하는 메서드를 분리된 객체를 사용하여
수정하였다.

parker1609 committed:
refactor: LottoNumber 객체로 변경

* 모델 객체 사이에서는 최대한 원시값을 포장한 LottoNumber 객체로
통신하도록 변경하였다.

parker1609 committed:
refactor: 보너스볼이 일치하는지 여부 추가

* Rank.java: enum 생성자에 보너스볼이 일치하는지 여부를 나타내는
boolean값을 추가하였다.

parker1609 committed:
refactor: 오류시 정상적으로 재입력

* 수동 로또 입력 때 정상적인 입력 후 비정상 입력을 할 경우 정상적인 입력이 남아있는 현상을 List<Lotto> 변수를 지역 변수로 변경하여 해결하였다.

parker1609 committed:
refactor: 자료형 변경

* Integer 자료형의 범위를 넘어선 계산을 하기 위해 BigInteger 클래스를 사용하였다.

parker1609 committed:
feat: 로또 개수를 담당하는 객체 추가

* LottoCount.java: 로또 개수를 담당하는 객체, 전체 로또 개수보다 많은 수동 로또 개수를 입력하면 오류를 발생시킨다.

* Money.java: isOverTotalNumberOfLotto() 메서드는 매개변수로 전달받은 로또 개수가 전체 로또 개수보다 크면 true 를 반환한다.

* LottoCountTest.java: 전체 로또 개수보다 큰 수동 로또 개수로 LottoCount 객체를 만들 때 오류가 정상적으로 발생하는지 테스트한다.

parker1609 committed:
feat: 수동 로또 개수 반환 메서드 생성

parker1609 committed:
feat: 자동 로또 개수 반환 메서드 생성

parker1609 committed:
refactor: LottoCount 객체 적용

parker1609 committed:
refactor: 사용하지 않는 부분 삭제

parker1609 committed:
refactor: 예외 케이스 해결

* 2, 3등을 제외하고는 보너스볼이 일치하는 지 여부와 상관없기 때문에 이를
처리하는 로직을 추가하였다.

parker1609 commented:
로또 개수를 담당하는 LottoCount 클래스를 추가하였습니다.

parker1609 committed:
READMD.md 정리

jihan805 commented on src/main/java/lotto/Main.java:
```diff
@@ -0,0 +1,88 @@
+        }
+    }
+
+    private static List<Lotto> createLottos(int numberOfManualLotto, int numberOfAutoLotto) {
```
공통적인 요소를 추출하여 LottoMachine 인터페이스를 구현한 부분 💯 
Main을 조금 더 간소화할 수 있을 것 같아요.
수동/자동 로또 생성을 합하여 완전한 로또를 생성해주는 클래스를 만들어보면 어떨까요?
즉, createLottos를 따로 추출하면 어떨까요?

해당 부분 구현후 이번단계 머지 진행하도록 하겠습니다!

parker1609 replied:
피드백 감사합니다!

Lottos의 생성을 담당하는 LottosFactory 클래스를 추가하였습니다. 그 결과, Main에서 생성을 담당하는 메서드 2개 정도 줄일 수 있었고, main() 메서드 내부도 조금 더 이해하기 쉬운 코드로 바뀔 수 있었습니다.

jihan805 commented on src/main/java/lotto/Main.java:
```diff
@@ -0,0 +1,88 @@
+        OutputConsole.outputLotto(lottos, lottoCount.getManualCount(), lottoCount.getAutoCount());
+        WinningLotto winningLotto = createWinningLotto(createLastWinningLotto());
+        OutputConsole.outputResult(new LottoResult(winningLotto, lottos));
+    }
```
전체적으로 구현 👍 
정말 잘하셨어요.
가독성을 위해 비슷한 개념을 하나의 덩어리로 묶고, 이를 기준으로 한줄 띄어보면 어떨까요?

parker1609 replied:
피드백 감사합니다!

제가 생각하기에 비슷하다고 생각되는 로또 생성, 로또 결과 및 출력으로 구분하였습니다.

jihan805 reviewed (changes requested):
안녕하세요 코맥님, 
처음에 너무 잘 구현하셔서 피드백도 잘 반영해주신것 같아요 👍 
정말 간단한 피드백 한개 남겼는데 확인 부탁드릴게요 :)

parker1609 committed:
README.md 수정

parker1609 committed:
refactor: 로또 생성 클래스 추출

* LottosFactory.java: 로또 개수와 수동 로또 번호의 정보를 가지고 전체
로또를 만드는 객체이다.

parker1609 committed:
refactor

* OutputConsole.java: outputLotto()메서드에서 불필요한 인수를 줄였다.

* Main.java: main()메서드는 전체적인 로또 프로그램을 수행하는데, 비슷한 수행끼리 구분하기 위해 줄바꿈을 추가하였다.

parker1609 committed:
refactor: 메서드 이름 변경

parker1609 committed:
refactor: 비슷한 개념끼리 묶기

parker1609 committed:
refactor: 인터페이스 메서드 수정

* LottoMachine.java: getLottos() 메서드의 반환 값을 List<Lotto>로
변경하였다.

* AutoLottoMachine.java: 생성자로 자동 로또 개수를 전달받아 그 만큼의
로또를 생성한다.

* ManualLottoMachine.java: 생성자로 로또 번호 6개의 리스트를 전달 받아
그 만큼의 로또를 생성한다.

* ConvertLottoNumber.java: List<Integer>를 List<LottoNumber>로 변환한다.

jihan805 commented on src/main/java/lotto/Main.java:
```diff
@@ -0,0 +1,83 @@
+        }
+    }
+
+    private static List<List<Integer>> createManualLottoNumbers(int manualCount)
```
List<<List>>와 같은 구조는 피하면 좋을 것 같아요!
수동으로 구매할 번호를 입력해달라는 메시지를 InputView로 옮기고, 파라미터로 몇개 입력받을지에 대한 count를 넘겨, 
String의 리스트로 받으면 수동 로또를 위한 인풋 부분이 조금 더 간결해질 수 있을 것 같아요!

jihan805 commented on src/main/java/lotto/domain/LottosFactory.java:
```diff
@@ -0,0 +1,37 @@
+    private final List<List<Integer>> manualLottoNumbers;
+    private final LottoCount lottoCount;
+    private List<Lotto> lottosList = new ArrayList<>();
+
```
LottosFactory 클래스를 추출하여 전체 로또 생성 👍 
변수를 선언하지 않고도 충분히 구현할 수 있을 것 같아요!
추가로 Main  에도 피드백 남겼듯 List<List<Integer>>의 구조를 개선할 수 있을 것 같아요! List 안에 List를 사용하는 이중 List 구조를 List 하나만 사용하도록 변경해보면 어떨까요?

jihan805 reviewed (approved):
안녕하세요 코맥님, 피드백 반영 잘해주셨어요 👍 
간단한 피드백 남겼으니, 다음 단계 진행하면서 고민부탁드려요.
머지 진행할게요 :)
